File: 0965) Univalued Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isUnivalTree(TreeNode* root) {
        if(root == nullptr)
        {
            return true;
        }
        int k = root->val;
        return check(root, k);
    }
    bool check(TreeNode *root, int k)
    {
        if(root == nullptr)
        {
            return true;
        }
        if(root->val != k)
        {
            return false;
        }
        return (check(root->left, k) && check(root->right, k));
    }
};

File: 0965) Univalued Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isUnivalTree(TreeNode root) {
        if(root == null)
        {
            return true;
        }
        int k = root.val;
        return helper(root, k);
    }
    public boolean helper(TreeNode root, int k)
    {
        if(root == null)
        {
            return true;
        }
        if(root.val != k)
        {
            return false;
        }
        return helper(root.left, k) && helper(root.right, k);
    }
}
